homeassistant/components/sensor/teleinfo.py

```diff
@@ -0,0 +1,149 @@
+It provides real time information on power consumption, rates and current on
+a user accessible serial port.
+
+Work based on https://github.com/nlamirault
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.teleinfo/
+
+Sample configuration.yaml
+
+    sensor:
+    - platform: teleinfo
+        name: "EDF teleinfo"
+        serial_port: "/dev/ttyAMA0"
+    - platform: template
+        sensors:
+        teleinfo_base:
+            value_template: '{{ (states.sensor.edf_teleinfo.
+                attributes["BASE"] | float / 1000) | round(0) }}'
+            unit_of_measurement: 'kWh'
+            icon_template: mdi:flash
+    - platform: template
+        sensors:
+        teleinfo_iinst1:
+            value_template: '{{ states.sensor.edf_teleinfo.
+                attributes["IINST1"] | int }}'
+            unit_of_measurement: 'A'
+            icon_template: mdi:flash
+
+"""
+import logging
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_NAME, EVENT_HOMEASSISTANT_STOP)
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['pyserial-asyncio==0.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_SERIAL_PORT = 'serial_port'
+
+CONF_ATTRIBUTION = "Provided by EDF Teleinfo."
+
+DEFAULT_NAME = "Serial Teleinfo Sensor"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SERIAL_PORT): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the Serial sensor platform."""
+    name = config.get(CONF_NAME)
+    port = config.get(CONF_SERIAL_PORT)
+    sensor = SerialTeleinfoSensor(name, port)
+
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, sensor.stop_serial_read())
+    async_add_entities([sensor])
+
+
+class SerialTeleinfoSensor(Entity):
+    """Representation of a Serial sensor."""
+
+    def __init__(self, name, port):
+        """Initialize the Serial sensor."""
+        self._name = name
+        self._port = port
+        self._serial_loop_task = None
+        self._state = None
+        self._attributes = {}
+
+    async def async_added_to_hass(self):
+        """Handle when an entity is about to be added to Home Assistant."""
+        self._serial_loop_task = self.hass.loop.create_task(
+            self.serial_read(self._port, baudrate=1200, bytesize=7,
+                             parity='E', stopbits=1, rtscts=1))
+
+    async def serial_read(self, device, **kwargs):
+        """Process the serial data."""
+        import serial_asyncio
+        _LOGGER.debug(u"Initializing Teleinfo")
+        reader, _ = await serial_asyncio.open_serial_connection(url=device,
```

Comment:
The device interface code I/O and data parsing should be moved to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Reply:
Hi, Martin this code is the untouched part of 
https://www.home-assistant.io/components/sensor.serial/
https://github.com/home-assistant/home-assistant/blob/master/homeassistant/components/sensor/serial.py
So i don't understand the request as you have already this code accepted in HA...
I don't have time to makes this changes on this platform.
But if you do the job with sensor.serial, I'll surely adapt my PR.

Reply:
Old code doesn't mean that it's a good example. We're very strict on the mentioed rule now. We haven't always been.

So we won't merge this PR until the interface code is moved to a standalone library.

Reply:
Guess that could be used : https://github.com/demikl/python-teleinfo